File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.java
import java.util.*;

class Solution {
    static class Ratio implements Comparable<Ratio> {
        int x, y;
        double gain;

        Ratio(int x, int y) {
            this.x = x;
            this.y = y;
            this.gain = (double) (x + 1) / (y + 1) - (double) x / y;
        }

        @Override
        public int compareTo(Ratio other) {
            return Double.compare(other.gain, this.gain);
        }
    }

    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<Ratio> pq = new PriorityQueue<>();
        for (int[] cls : classes) {
            pq.add(new Ratio(cls[0], cls[1]));
        }

        for (int i = 0; i < extraStudents; i++) {
            Ratio top = pq.poll();
            top.x++;
            top.y++;
            top.gain = (double) (top.x + 1) / (top.y + 1) - (double) top.x / top.y;
            pq.add(top);
        }

        double totalRatio = 0;
        while (!pq.isEmpty()) {
            Ratio top = pq.poll();
            totalRatio += (double) top.x / top.y;
        }

        return totalRatio / classes.length;
    }
}

File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.cpp
class Solution {
public:
    struct ratio{
        int x,y;
        double r=(1.0*(x+1) / (y+1) ) - (1.0*x / y);
        bool operator<(const ratio &other)const{
            return other.r > r;
        }
    };
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        priority_queue<ratio> pr;
        for(auto it : classes)pr.push({it[0],it[1]});
       
       while(extraStudents--){
        auto t = pr.top();
        t.x++;t.y++; pr.pop();
        pr.push({t.x,t.y});
       }
       double ans = 0;
       while(pr.size()){
        int x = pr.top().x;
        int y = pr.top().y;
        pr.pop();
        ans+=(1.0*x/y);
       }
       return ans / classes.size();
    }
};